dotnet/src/Connectors/Connectors.OpenAI/Core/OpenAIFunction.cs

```diff
@@ -127,52 +139,143 @@ internal OpenAIFunction(
+                });
         }
 
         return ChatTool.CreateFunctionTool
         (
             functionName: this.FullyQualifiedName,
             functionDescription: this.Description,
-            functionParameters: resultParameters
+            functionParameters: resultParameters,
+            functionSchemaIsStrict: allowStrictSchemaAdherence
         );
     }
 
     /// <summary>Gets a <see cref="KernelJsonSchema"/> for a typeless parameter with the specified description, defaulting to typeof(string)</summary>
-    private static KernelJsonSchema GetDefaultSchemaForTypelessParameter(string? description)
+    private static KernelJsonSchema GetDefaultSchemaForTypelessParameter(string? description, bool allowStrictSchemaAdherence)
     {
         // If there's a description, incorporate it.
         if (!string.IsNullOrWhiteSpace(description))
         {
-            return KernelJsonSchemaBuilder.Build(typeof(string), description);
+            return allowStrictSchemaAdherence ?
+                GetOptionalStringSchemaWithDescription(description!) :
+                KernelJsonSchemaBuilder.Build(typeof(string), description, AIJsonSchemaCreateOptions.Default);
         }
 
         // Otherwise, we can use a cached schema for a string with no description.
-        return s_stringNoDescriptionSchema;
+        return allowStrictSchemaAdherence ? s_stringNoDescriptionSchemaAndNull : s_stringNoDescriptionSchema;
+    }
+
+    private static KernelJsonSchema GetOptionalStringSchemaWithDescription(string description)
+    {
+        var jObject = new JsonObject
+        {
+            { "description", description },
+            { "type", new JsonArray { "string", "null" } },
+        };
+        return KernelJsonSchema.Parse(jObject.ToString());
+    }
+    private static KernelJsonSchema? GetSanitizedSchemaForStrictMode(KernelJsonSchema? schema, bool insertNullType)
+    {
+        if (schema is null)
+        {
+            return null;
+        }
+        var forbiddenPropertyNames = s_forbiddenKeywords.Where(k => schema.RootElement.TryGetProperty(k, out _)).ToArray();
+
+        if (forbiddenPropertyNames.Length > 0 || insertNullType && schema.RootElement.TryGetProperty(TypeKey, out var typeElement) &&
+            (typeElement.ValueKind == JsonValueKind.Array && !typeElement.EnumerateArray().Any(static t => NullType.Equals(t.GetString(), StringComparison.OrdinalIgnoreCase)) ||
+            typeElement.ValueKind == JsonValueKind.String && typeElement.GetString() != NullType))
+        {
+            var originalSchema = JsonSerializer.Serialize(schema.RootElement);
+            var parsedJson = JsonNode.Parse(originalSchema);
+
+            if (parsedJson is null)
+            {
+                return schema;
+            }
+
+            var jsonObject = parsedJson.AsObject();
+            foreach (var forbiddenPropertyName in forbiddenPropertyNames)
+            {
+                jsonObject.Remove(forbiddenPropertyName);
```

Comment:
this is where I'm doing the forbidden keywords cleanup, might want to be recursive in case we have those on the request body.
more context https://github.com/microsoft/semantic-kernel/pull/9807#discussion_r1873817894